design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+- "SDK" or Software Development Kit usually refers to the rendering library, like OpenGL or Metal, at least in this document.
+- "API" refers to an Application Programming Interface.  In this context it refers to the classes and methods made available to a developer.
+
+## Goals
+
+It is useful to split our goals into three sections to articulate what this proposal aims to accomplish:
+
+- **North Stars:** These guide the overall direction of the proposal, and are the overarching _why_.
+
+- **Core Functionality:** These are concrete changes we are proposing to the library to get to our North Stars.
+
+- **Evaluation metrics:** These are specific criteria we will accomplish through this proposal.
+
+### <a name="north-stars">North Stars</a>
+1. The MapLibre Native toolkit should use Metal for rendering on iOS
+2. The MapLibre Native toolkit should use Metal for rendering on MacOS
+
+### <a name="core">Core Functionality</a>
+1. The OpenGL renderer will still be available on iOS as a compile time flag
+2. Shaders will be reimplemented in Metal
+3. Off screen targets will be implemented in Metal and exposed in a such a way they can be read from and shared
+4. Rendering passes will be implemented in Metal
+5. Snapshots will not block the rendering pipeline
+6. Atlases will be implemented for Metal and be accessible from any thread
+7. The Metal renderer will be optimized to the best of our abilities
+8. The toolkit will allow mixing of real time assets from other Metal based toolkits 
+
+###  <a name="eval">Evaluation Metrics</a>
+1. The new version of MapLibre with Metal should support all iOS devices that support Metal
+2. .... should not be any slower than the OpenGL version
+3. .... should be significantly faster in most test cases
+4. .... should use the main thread less than the previous version on iOS
+
+## Proposed Changes
+
+Before we dive into the specific changes we wish to make, let's discuss how software development works with one of these low level rendering SDKs.
+
+Anyone who has worked with OpenGL knows that first you get nothing.  Traditionally, it's a black screen and a lot of tweaking before the developer sees anything.  And then a lot more tweaking before the developer sees anything useful.  
+
+With Metal, you get crashes with error messages usually related to what you actually did.  This is a huge improvement!  And the debugging tools... well they do exist and they mostly work.  They work better in direct mode and with smaller test cases, so we plan accordingly.  When they do work, you can actually debug a shader!  That's very helpful.
+
+A Metal implementation is conceptually easier than OpenGL, but it's still real time rendering and you can still do a variety of frustrating things that are a difficult to debug (like crash the hardware).  Developers will run into weird cases that are hard to debug and you need to plan to lose a few days here and there.
```

Comment:
> Since we are talking about user-space APIs, there should be no way to crash the hardware

You're right.  There *should* be no way to crash the hardware.  But if you're looking for one... my favorite is a shader with memory issues.

The conceptual issue with OpenGL vs Metal is this.  OpenGL doesn't describe how the GPU works and hasn't in quite some time.  That makes getting good performance much harder.  Of course, the consortium and its contributors and partners knew this so they added extensions (so many extensions).  Those extensions, collectively, represent the modern GPU a bit better, but they're weird to use.

So if you start with a naive implementation which is fairly simple (though OpenGL is not that simple) you'll need to add in various extensions to get closer to what the GPU actually wants.  That gets fairly odd and difficult to understand.  With modern toolkits, Metal included, you start out with what the GPU is expecting and work from there.  

That's what I mean by conceptually easier.

Reply:
I think we agree!  I'll consider wording.

Reply:
Reworded